tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java

```diff
@@ -321,6 +322,103 @@ public static void test_multiple_connections() throws Exception {
 		stmt1.close();
 	}
 
+    public static void test_duckdb_timestamp() throws Exception {
+		Connection conn = DriverManager.getConnection("jdbc:duckdb:");
+		Statement stmt = conn.createStatement();
+		stmt.execute("CREATE TABLE a (ts TIMESTAMP)");
+
+        // Generat tests without database
+        Timestamp ts0 = Timestamp.valueOf("1970-01-01 00:00:00");
+        Timestamp ts1 = Timestamp.valueOf("2021-07-29 21:13:11");
+        Timestamp ts2 = Timestamp.valueOf("2021-07-29 21:13:11.123456");
+
+        Timestamp cts0 = new DuckDBTimestamp(ts0).toSqlTimestamp();
+        Timestamp cts1 = new DuckDBTimestamp(ts1).toSqlTimestamp();
+        Timestamp cts2 = new DuckDBTimestamp(ts2).toSqlTimestamp();
+        
+        assertTrue(ts0.getTime() == cts0.getTime());
+        assertTrue(ts0.compareTo(cts0) == 0);
+        assertTrue(ts1.getTime() == cts1.getTime());
+        assertTrue(ts1.compareTo(cts1) == 0);
+        assertTrue(ts2.getTime() == cts2.getTime());
+        assertTrue(ts2.compareTo(cts2) == 0);
+
+        assertTrue(DuckDBTimestamp.getMicroseconds(DuckDBTimestamp.toSqlTimestamp(5678912345L)) == 5678912345L);
+
+        DuckDBTimestamp dts4 = new DuckDBTimestamp(ts1);
+        assertTrue(dts4.toSqlTimestamp().compareTo(ts1) == 0);
+        DuckDBTimestamp dts5 = new DuckDBTimestamp(ts2);
+        assertTrue(dts5.toSqlTimestamp().compareTo(ts2) == 0);
+
+        // Insert and read a timestamp 
+        stmt.execute("INSERT INTO a (ts) VALUES ('2005-11-02 07:59:58')");
+		ResultSet rs = stmt.executeQuery(
+				"SELECT * FROM a");
+		assertTrue(rs.next());
+		assertEquals(rs.getObject("ts"), Timestamp.valueOf("2005-11-02 07:59:58"));
+		assertEquals(rs.getTimestamp("ts"), Timestamp.valueOf("2005-11-02 07:59:58"));
+    
+		rs.close();
+		stmt.close();
+
+		PreparedStatement ps = conn.prepareStatement(
+				"SELECT COUNT(ts) FROM a WHERE ts = ?");
+		ps.setTimestamp(1, Timestamp.valueOf("2005-11-02 07:59:58"));
+        ResultSet rs2 = ps.executeQuery();
+		assertTrue(rs2.next());
+        assertEquals(rs2.getInt(1), 1);
+        rs2.close();
+        ps.close();
+
+		ps = conn.prepareStatement(
+				"SELECT COUNT(ts) FROM a WHERE ts = ?");
+		ps.setObject(1, Timestamp.valueOf("2005-11-02 07:59:58"));
+        ResultSet rs3 = ps.executeQuery();
+		assertTrue(rs3.next());
+        assertEquals(rs3.getInt(1), 1);
+        rs3.close();
+        ps.close();
+
+		ps = conn.prepareStatement(
+				"SELECT COUNT(ts) FROM a WHERE ts = ?");
+		ps.setObject(1, Timestamp.valueOf("2005-11-02 07:59:58"), Types.TIMESTAMP);
+        ResultSet rs4 = ps.executeQuery();
```

Comment:
This is where you could check the local/GMT time values. The `EPOCH` function will tell you what the count of seconds from the GMT epoch is for each value.

Reply:
Maybe I missed something, but comparing the seconds from epoch between DuckDB's timestamp and Java's timestamp will not work if the client has not GMT. Only the result of the toString() functions can be compared and must be equal in the end.

Reply:
What exactly does Java store in these objects? I was under the impression it was an instant that got binned in the local time zone, but it sounds like they are actually storing naïve timestamps?

Reply:
To sum it up: 
The older Java objects (Timestamp, Date) may store the diff to epoch and the timezone. The complete replacement with the new java.time package indicate that there were problems with the implementation and approach.

The newer LocalDateTime has a LocalDate and LocalTime, which keep the year, months, (...) as ints. This means they are unaware of the timezone.
The newer OffsetDateTime stores a LocalDateTime and a timezone offset.